cns/nodesubnet/ip_fetcher.go

```diff
@@ -7,49 +7,144 @@ import (
 	GetInterfaceIPInfo(ctx context.Context) (nmagent.Interfaces, error)
 }
 
-type IPFetcher struct {
-	// Node subnet state
-	secondaryIPQueryInterval   time.Duration // Minimum time between secondary IP fetches
-	secondaryIPLastRefreshTime time.Time     // Time of last secondary IP fetch
+// IPConsumer is an interface implemented by whoever consumes the secondary IPs fetched in nodesubnet
+type IPConsumer interface {
+	UpdateIPsForNodeSubnet([]netip.Addr) error
+}
 
-	ipFectcherClient InterfaceRetriever
+// IPFetcher fetches secondary IPs from NMAgent at regular intervals. The
+// interval will vary within the range of minRefreshInterval and
+// maxRefreshInterval. When no diff is observed after a fetch, the interval
+// doubles (subject to the maximum interval). When a diff is observed, the
+// interval resets to the minimum.
+type IPFetcher struct {
+	// Node subnet config
+	intfFetcherClient  InterfaceRetriever
+	ticker             refreshticker.TickProvider
+	tickerInterval     time.Duration
+	consumer           IPConsumer
+	minRefreshInterval time.Duration
+	maxRefreshInterval time.Duration
 }
 
-func NewIPFetcher(nmaClient InterfaceRetriever, queryInterval time.Duration) *IPFetcher {
+// NewIPFetcher creates a new IPFetcher. If minInterval is 0, it will default to 4 seconds.
+// If maxInterval is 0, it will default to 1024 seconds (or minInterval, if it is higher).
+func NewIPFetcher(
+	client InterfaceRetriever,
+	consumer IPConsumer,
+	minInterval time.Duration,
+	maxInterval time.Duration,
+) *IPFetcher {
+	if minInterval == 0 {
+		minInterval = DefaultMinRefreshInterval
+	}
+
+	if maxInterval == 0 {
+		maxInterval = DefaultMaxRefreshInterval
+	}
+
+	maxInterval = max(maxInterval, minInterval)
+
 	return &IPFetcher{
-		ipFectcherClient:         nmaClient,
-		secondaryIPQueryInterval: queryInterval,
+		intfFetcherClient:  client,
+		consumer:           consumer,
+		minRefreshInterval: minInterval,
+		maxRefreshInterval: maxInterval,
+		tickerInterval:     minInterval,
+	}
+}
+
+// UpdateFetchIntervalForNoObservedDiff informs IPFetcher that no diff was observed in the last fetch.
+// In the current design, this doubles the fetch interval, subject to the maximum interval.
+func (c *IPFetcher) UpdateFetchIntervalForNoObservedDiff() {
+	c.tickerInterval = min(c.tickerInterval*2, c.maxRefreshInterval) //nolint:gomnd // doubling interval
+
+	if c.ticker != nil {
+		c.ticker.Reset(c.tickerInterval)
 	}
 }
 
-func (c *IPFetcher) RefreshSecondaryIPsIfNeeded(ctx context.Context) (ips []netip.Addr, err error) {
-	// If secondaryIPQueryInterval has elapsed since the last fetch, fetch secondary IPs
-	if time.Since(c.secondaryIPLastRefreshTime) < c.secondaryIPQueryInterval {
-		return nil, ErrRefreshSkipped
+// UpdateFetchIntervalForNoObservedDiff informs IPFetcher that a diff was observed in the last fetch.
+// In the current design, this resets the fetch interval to the minimum.
+func (c *IPFetcher) UpdateFetchIntervalForObservedDiff() {
```

Comment:
I don't like that controlling the fetch interval is controllable outside of `*IPFetcher`. This feels like a leaky abstraction.